Jamie/memory management leak fixes

jbeverly opened the pull request:
Overhaul of the glpk memory management environment code to allow state-object which is migrated before/after each glpk function on any given thread (so all allocation tracking is still done on TLS env, but we migrate everything that happened on that TLS env to a "master" env which is per-problem/mathprog/tree object (similar to `_r` API, but using the existing ENV for lockfree allocation.))

The structure here, is to use glpk's existing `env` code (which is a thread-local store for per-env data management) in a way that works in the Worker model of v8. Where each call to a problem object could (and usually does) land on a different thread. Since each call is, itself "pinned" to a thread, we can migrate a "common" env (env_state_ in the problem object) to the TLS env; let the normal glpk allocation and memory management operations happen on that TLS, and then when we're done, migrate everything that happened on that TLS back to "common" env.

Since reads/writes from the common env must be synchronized, that env is nested in a state struct object with a rwlock on it.

Note that this PR undoes some of the work I had done previously, since this fixes more than that work had fixed, and does it in a better way (and eliminates memory-leaks due to thread-switching between workers and threads being joined/detached causing the TLS on those workers to be "lost" without having been freed)

jbeverly committed:
add MBD back, will migrate to mutex protected LL on thread death

jbeverly committed:
add global env api

jbeverly committed:
use state object pass to (more) reentrant API; still using env

jbeverly committed:
preserve counters after Delete method

jbeverly committed:
fix test, and typo in problem.hpp

jbeverly committed:
missed a file on the prior commit

erikbryant commented on src/glpk/env/alloc.c:
```diff
@@ -45,74 +46,141 @@
+ * Free an mbd from an env. If the mbd's env and the current env mismatch, the block will *not* be deallocated yet,
+ * instead the deallocation will be deferred until the environment is freed
+ */
+static inline void* _free_mbd(MBD* mbd, ENV* env) {
```
I'm curious about the use of 'inline'. I thought that most modern compilers were smart enough to figure out whether to inline or not automatically.

jbeverly replied:
It serves 2 main purposes 1) It informs the compiler of my intent for it to be inline, and so perhaps increases the odds of that actually happening. and 2) It communicates the intent to readers of this code that they should not take the address of it, as that would prevent it from being inlined. 

Incidentally, the compiler may still choose not to inline it. You have to jump through more (compiler-specific) hoops to force the compiler to inline it. (for example in gcc `static inline __attribute__((__always_inline__))` will  at least cause taking its address to generate a warning with `-Winline`)

erikbryant commented on src/glpk/env/alloc.c:
```diff
@@ -45,74 +46,141 @@
+
+static inline void _check_allocation(ENV* env, size_t size) 
+{
+    if (size > SIZE_T_MAX - MBD_SIZE)
```
Much of the rest of the code uses braces for even single-line if statements. What about adding braces here?

jbeverly replied:
Nice catch; I thought I'd added them everywhere I touched; but since this was moved I missed it.

jbeverly replied:
fixed in 9281f64

erikbryant commented on src/glpk/env/alloc.c:
```diff
@@ -45,74 +46,141 @@
+
+static inline void _prepend_mbd_to_env(MBD* mbd, ENV* env, size_t size)
+{
+    xassert(mbd != NULL && env != NULL);
```
Splitting this into two separate asserts can help in debugging. If the conditions are combined in a single assert, you will not know which one of the conditions failed.

jbeverly replied:
fixed in 9281f64

erikbryant commented on src/glpk/env/alloc.c:
```diff
@@ -45,74 +46,141 @@
+        xerror("too many memory blocks allocated\n");
+}
+
+static inline void _prepend_mbd_to_env(MBD* mbd, ENV* env, size_t size)
```
Can you add some comments? It is clear what mdb and env are, but I'm not sure what size is for. I'm guessing it is the size of the env, but I'm not sure.

jbeverly replied:
fixed in 9281f64

erikbryant commented on src/glpk/env/alloc.c:
```diff
@@ -45,74 +46,141 @@
+}
+
+/**
+ * Reallocate an mdb. This can *only* work if the mbd was previously allocated in this env. 
```
```suggestion
 * Reallocate an mbd. This can *only* work if the mbd was previously allocated in this env. 
```

erikbryant commented on src/glpk/env/alloc.c:
```diff
@@ -45,74 +46,141 @@
+        return _free_mbd(mbd, env);
+    }
+    if(mbd->env != env) {
+        xerror("%s: ptr = %p; unable to reallocate from another environment", mbd);
```
I see two formatting directives, but only one parameter. Is one missing?

jbeverly replied:
fixed in 9281f64

erikbryant commented on src/glpk/env/alloc.c:
```diff
@@ -45,74 +46,141 @@
+    }
+    _check_allocation(env, size);
+    _remove_from_env(mbd, env);
+    xassert((mbd = realloc(mbd, size + MBD_SIZE)) != NULL);
```
```suggestion
        mbd = realloc(mbd, size + MBD_SIZE);
        xassert(mdb != NULL);
```

erikbryant replied:
I think this is easier to read if it is split into two statements. Also, I've seen compiler directives that will #define away assert statements. Not sure if this library supports that, but if it does then the call to realloc() will get discarded.

erikbryant commented on src/glpk/env/alloc.c:
```diff
@@ -45,74 +46,141 @@
+        return _free_mbd(mbd, env);
+    }
+    if(mbd->env != env) {
+        xerror("%s: ptr = %p; unable to reallocate from another environment", mbd);
```
Most other calls to xerror() include a trailing\n'. Should this one?

jbeverly replied:
fixed in 9281f64

erikbryant commented on src/glpk/env/alloc.c:
```diff
@@ -248,7 +316,7 @@ void glp_mem_limit(int limit)
 {  
-#ifndef GLOBAL_MEM_STATS
+#ifdef HAVE_ENV
       ENV *env = get_env_ptr();
```
I don't see that env is used after it is defined.

erikbryant commented on src/glpk/env/env.c:
```diff
@@ -233,43 +243,149 @@ const char *glp_version(void)
+{
+    return pthread_rwlock_wrlock(&env_state->env_lock);
+}
+/** Free lock
```
```suggestion
/**
 * Free lock
```

erikbryant commented on src/glpk/env/env.c:
```diff
@@ -233,43 +243,149 @@ const char *glp_version(void)
+/**
+ * Get env lock
+ */
+static inline int environ_state_wrlock(glp_environ_state_t *env_state)
```
```suggestion

static inline int environ_state_wrlock(glp_environ_state_t *env_state)
```

erikbryant commented on src/glpk/env/env.c:
```diff
@@ -233,43 +243,149 @@ const char *glp_version(void)
+{
+    return pthread_rwlock_rdlock(&env_state->env_lock);
+}
+/**
```
```suggestion

/**
```

erikbryant commented on src/glpk/env/env.c:
```diff
@@ -233,43 +243,149 @@ const char *glp_version(void)
+
+void glp_env_tls_init_r(glp_environ_state_t *env_state, void* info) 
+{
+ //   glp_init_env();
```
Delete this line?

jbeverly replied:
fixed in 9281f64

erikbryant commented on src/common.h:
```diff
@@ -43,8 +43,7 @@ Nan::DefineOwnProperty(target, constant_name, constant_value, constant_attribute
 #define GLP_CREATE_HOOK_GUARDS(NAME) \
-    MemStatsGuard mguard{NAME->memstats_}; \
-    TermHookGuard hookguard{NAME->info_}
+    GLPKEnvStateGuard mguard{NAME->env_state_, NAME->info_}; \
```
```suggestion
    GLPKEnvStateGuard mguard{(NAME)->env_state_, (NAME)->info_}; \
```

erikbryant replied:
I think the recommended practice is to put parens around any of the #define parameters. This prevents parsing issues if the parameter is actually an expression. For instance, I believe this would fail:

`GLP_CREATE_HOOK_GUARDS( (char *)ptr - MBD_SIZE )

erikbryant commented on src/glpk/env/alloc.c:
```diff
@@ -45,74 +46,141 @@
 #ifdef HAVE_ENV
+
+/**
+ * "Unwrap" the memory block descriptor
```
```suggestion
 * Return the memory block descriptor for a memory block that was allocated by glp_alloc.
```

erikbryant commented on src/glpk/env/alloc.c:
```diff
@@ -45,74 +46,141 @@
+    _remove_from_env(mbd, env);
+    free(mbd);
+    return NULL;
+}
```
If you are concerned about performance/memory, I think this would be smaller:

```
    if(mbd->env == env) {
        _remove_from_env(mbd, env);
        free(mbd);
    }
    return NULL;
```

jbeverly replied:
fixed in 9281f64

erikbryant commented on src/glpk/env/alloc.c:
```diff
@@ -45,74 +46,141 @@
+    }
+    _remove_from_env(mbd, env);
+    free(mbd);
+    return NULL;
```
It is not clear why the function needs to return a value. Could it be changed to not return one?

jbeverly replied:
clarified by making function-like macros into more obvious MACROS in 9281f64

mcwitt commented:
While the PR description covers details of the implementation, I think it'd be really helpful for reviewers if you could give an overview of the specific problem that existed before this and how this PR solves it. I gather from offline conversation and the last sentence of the PR description that the change here

> eliminates memory-leaks due to thread-switching between workers and threads being joined/detached causing the TLS on those workers to be "lost" without having been freed

I only have a vague sense of how this PR solves that particular problem; my current interpretation is something like "V8 kills worker threads without giving them a chance to clean up allocated resources, so deallocation has to be done in the master thread, and the workers have to update some common state so that the master knows what needs to be cleaned up." Is this roughly correct? Even so, I think it'd be useful if you could provide a bit more detail here.

Also, for interested reviewers not familiar with V8 and/or GLPK (asking for a friend :P), I think a brief glossary of the following terms would be very useful:
- TLS env (is this a Worker's env?)
- `_r` API
- lockfree allocation
- problem/mathprog/tree object

mcwitt commented on src/glpk/env/alloc.c:
```diff
@@ -45,74 +46,141 @@
+
+    /* update counters */
+    if (get_mem_count() < 1 || get_mem_total() < mbd->size) {
+        xerror("unlinking mbd: memory allocation error\n");
```
Seems like `"inconsistent state"` could be a more appropriate error message here, since it seems like the trigger would be trying to deallocate when state says there is nothing allocated.

jbeverly replied:
fixed in 9281f64

mcwitt commented on src/glpk/env/alloc.c:
```diff
@@ -45,74 +46,141 @@
+        xerror("block too large\n");
+    if (size + MBD_SIZE > get_mem_limit() - get_mem_total())
+        xerror("memory allocation limit exceeded\n");
+    if (get_mem_count() == INT_MAX)
```
Should this be changed to `SIZE_T_MAX`? (it looks like `mem_count` is a `size_t`)

jbeverly replied:
fixed in 9281f64

erikbryant committed:
Avoid assignment in condition

(GNU coding convention, and just a good idea)

Co-Authored-By: jbeverly <jbeverly@users.noreply.github.com>

mcwitt commented on src/glpk/env/alloc.c:
```diff
@@ -45,74 +46,141 @@
+}
+
+/**
+ * Take the mbd off the environment. Must only ever try to remove an mdb from the environment
```
```suggestion
 * Take the mbd off the environment. Must only ever try to remove an mbd from the environment
```

erikbryant committed:
Typo

Co-Authored-By: jbeverly <jbeverly@users.noreply.github.com>

erikbryant committed:
Use parens in macro

Co-Authored-By: jbeverly <jbeverly@users.noreply.github.com>

erikbryant committed:
Better comment

Co-Authored-By: jbeverly <jbeverly@users.noreply.github.com>

erikbryant committed:
Vspace

Co-Authored-By: jbeverly <jbeverly@users.noreply.github.com>

erikbryant committed:
Comment in following line

Co-Authored-By: jbeverly <jbeverly@users.noreply.github.com>

mcwitt commented on src/glpk/env/alloc.c:
```diff
@@ -45,74 +46,142 @@
+    }
+    _check_allocation(env, size);
+    _remove_from_env(mbd, env);
+        mbd = realloc(mbd, size + MBD_SIZE);
```
Was the extra indent here intentional?

jbeverly replied:
fixed in 9281f64

mcwitt reviewed (commented):
Leaving a few minor comments for now. Still planning a more thorough review, but let's start testing.

jbeverly committed:
PR review sweep

- change function-like macros to obvious MACROS so variable usage is
  more obvious
- _free_mbd no longer returns NULL, and has single-exit point
- add braces around updated if blocks
- add comments explaining _prepend_mbd_to_env, glp_realloc,
  glp_env_tls_init_r, and glp_init_env_state
- fix xerror with too many format specifiers.
- fix 'mdb' several places.
- add newline to xerror

rregue commented:
@jbeverly I did what it is outlined on this PR: https://github.com/TakeScoop/node-glpk-playground/pull/2#issue-265138054 and wrote down some observations. The most relevant is that the memory stats are not being logged and that I was under the impression that things would break when running old glpk concurrently, but it seems that it worked.

rregue commented on test/memusage-test.js:
```diff
@@ -42,6 +43,7 @@ describe('Leakage check', function() {
                 let idx = 0;
                 lp.on('log', function(msg) {
                     let info = glp.glpMemInfo()
+                    console.log(`INDEX: ${idx++}: ${msg.trim()}`)
```
@jbeverly cannot comment below, but I logged the info object and it is all zeros. The test below will pass if everything is zero, so we should probably update. See:
```
INDEX: 22: +     3: mip =   7.333333333e+02 <=     tree is empty   0.0% (0; 1)
{
  "count": 0,
  "cpeak": 0,
  "total": 0,
  "tpeak": 0
}
INDEX: 23: INTEGER OPTIMAL SOLUTION FOUND
{
  "count": 0,
  "cpeak": 0,
  "total": 0,
  "tpeak": 0
}
```

rregue replied:
actually, I see a different behavior when logging in `problem-memusage-tests.js`

jbeverly committed:
rewire up glp.glpMemInfo; had forgotten it

jbeverly committed:
cosmetic changes and bugfix in linked-list merge
